File: leetcode/rust-leetcode/src/1657.determine-if-two-strings-are-close.rs
use std::collections::HashMap;
impl Solution {
    pub fn close_strings(word1: String, word2: String) -> bool {
        if word1.len() == word2.len() {
            let mut map = HashMap::new();
            let mut map2 = HashMap::new();
            for b in word1.bytes() {
                *map.entry(b).or_insert(0) += 1;
            }
            // println!("{:?}", map);
            for b in word2.bytes() {
                *map2.entry(b).or_insert(0) += 1;
            }
            // println!("{:?}", map2);
            if map.len() != map2.len() {
                return false;
            }
            for key in map.keys() {
                if !map2.contains_key(key) {
                    return false;
                }
            }
            let mut vals1 = map.values().copied().collect::<Vec<i32>>();
            vals1.sort();
            let mut vals2 = map2.values().copied().collect::<Vec<i32>>();
            vals2.sort();
            let matching = vals1.iter().zip(&vals2).filter(|&(a, b)| a == b).count();
            return matching == vals1.len()
        }
        false
    }
}

impl Solution {

    const TOTAL_LETTERS: usize = 'z' as usize + 1;

    pub fn close_strings(word1: String, word2: String) -> bool {
        if word1.len() != word2.len() {
            return false;
        }

        let mut total_letters1 = [0; Self::TOTAL_LETTERS];
        for ch in word1.chars() {
            total_letters1[ch as usize] += 1;
        }

        let mut total_letters2 = [0; Self::TOTAL_LETTERS];
        for ch in word2.chars() {
            let index = ch as usize;
            if total_letters1[index] == 0 {
                return false;
            }
            total_letters2[index] += 1;
        }

        total_letters1.sort_unstable();
        total_letters2.sort_unstable();

        total_letters1 == total_letters2
    }
}

File: leetcode/python/1657.determine-if-two-strings-are-close.py
class Solution:
    def closeStrings(self, word1: str, word2: str) -> bool:
        if len(word1) != len(word2):
            return False
        charCount = ord('z') + 1
        chars1 = [0] * charCount
        chars2 = [0] * charCount
        
        for c in word1:
            chars1[ord(c)] += 1
        
        for c in word2:
            if chars1[ord(c)] == 0:
                return False
            chars2[ord(c)] += 1
        
        chars1.sort()
        chars2.sort()

        return chars1 == chars2

""" """

class Solution:
    def closeStrings(self, word1: str, word2: str) -> bool:
        list1 =[] 
        list2 = [] 
        for i in set(word1):
            list1.append(word1.count(i)) 
            list2.append(word2.count(i))
        list1.sort()
        list2.sort()
        return list1 == list2

